apps/prairielearn/src/ee/models/plan-grants.ts

```diff
@@ -0,0 +1,98 @@
+import { loadSqlEquiv, queryRow } from '@prairielearn/postgres';
+
+import { insertAuditLog } from '../../lib/audit-log';
+import { PlanGrant, PlanGrantSchema, EnumPlanGrantType } from '../../lib/db-types';
+
+const sql = loadSqlEquiv(__filename);
+
+type NewBasePlanGrant = Omit<PlanGrant, 'created_at' | 'id'>;
+type NewInstitutionPlanGrant = Omit<
+  NewBasePlanGrant,
+  'course_instance_id' | 'enrollment_id' | 'user_id'
+>;
+type NewCourseInstancePlanGrant = Omit<NewBasePlanGrant, 'enrollment_id' | 'user_id'>;
+type NewEnrollmentPlanGrant = Omit<NewBasePlanGrant, 'user_id'>;
+type NewUserPlanGrant = Omit<
+  NewBasePlanGrant,
+  'institution_id' | 'course_instance_id' | 'enrollment_id'
+>;
+
+type NewPlanGrant =
+  | NewInstitutionPlanGrant
+  | NewCourseInstancePlanGrant
+  | NewEnrollmentPlanGrant
+  | NewUserPlanGrant;
+
+export async function insertPlanGrant(planGrant: NewPlanGrant): Promise<void> {
+  const newPlanGrant = await queryRow(
+    sql.insert_plan_grant,
+    {
+      institution_id: null,
+      course_instance_id: null,
+      enrollment_id: null,
+      user_id: null,
+      ...planGrant,
```

Comment:
I had a draft of that here: https://github.com/PrairieLearn/PrairieLearn/pull/5610. Might be nice to revisit that sometime!

Yeah, I can switch to passing them explicitly here.

Reply:
I got, uh, clever with types in 70e3487 and made this work 😅 let me know if this is too much!

Reply:
Hahaha, that's a pretty deep level of cleverness 😄. I don't follow all the details of how `WithRequiredKeys` is implemented but the usage of it looks clear and natural. I definitely like being able to avoid the splat.